torch/_inductor/lowering.py

```diff
@@ -6219,10 +6220,199 @@ def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs):
+            ),
+            num_init_leaves,
+        )
+        fake_ys_outs = [stack_y(y, fake_scan_length) for y in fake_ys_subgraph]
+        fake_idx = V.graph.fake_mode.shape_env.create_symintnode(  # type: ignore[union-attr]
+            V.graph.fake_mode.shape_env.create_unspecified_symbol(  # type: ignore[union-attr]
+                0, source=GlobalSource("scan_idx"), dynamic_dim=DimDynamic.DYNAMIC
+            ),
+            hint=0,
+            source=GlobalSource("scan_idx"),
+        )
+        fake_args, tree_spec = pytree.tree_flatten(
+            [fake_init, fake_xs, fake_additional_inputs, fake_ys_outs, fake_idx]
+        )
+
+        # Note: we lower scan into while_loop by constructing cond_fn and body_fn.
+        # We need the out variant graph because, for ys_outs, if we append them to a list then
+        # stack them together after the loop, we'll waste a lot of memory and it's also time
+        # consuming. We could optimize the overhead away by pre-allocating the memory.
+        # The cond_fn and body_fn are out variants, where ys_outs is passed in as a pre-allocated
+        # buffer and we pass in an additional SymInt to represent the current iteration idx.
+        # Tracing the graph produces a new graph, where the nodes depends on the value of symint
+        # to select from xs of input and copy into a slice of ys_outs.
+        def cond_fn_out(*flat_args):
+            init, xs, additional_inputs, ys_outs, idx = pytree.tree_unflatten(
+                flat_args, tree_spec
+            )
+            return idx < xs[0].size()[specialized_dim]
+
+        def body_fn_out(*flat_args):
+            init, xs, additional_inputs, ys_outs, idx = pytree.tree_unflatten(
+                flat_args, tree_spec
+            )
+            scan_idx = idx if not reverse else xs[0].size()[specialized_dim] - idx - 1
+            sub_xs = []
+            for x in xs:
+                sub_xs.append(torch.ops._scan_helper.unsafe_select(x, dim, scan_idx))
+
+            new_carry, ys = _extract_carry_and_out(
+                combine_subgraph.graph_module(
+                    *(list(init) + sub_xs + list(additional_inputs))
+                ),
+                num_init_leaves,
+            )
+            for y, y_out in zip(ys, ys_outs):
+                y_out_slice = torch.ops._scan_helper.unsafe_select(y_out, 0, scan_idx)
+                y_out_slice.copy_(y)
+            return *new_carry, *xs, *additional_inputs, *ys_outs
+
+        # Note:
+        # 1. fwd_only trace with real mode, we need to trace under current fake_mode otherwise
+        # factory functions will create real tensors instead of fake tensors.
+        # 2. We set run_functional_passes = False because body_fn_out is not functional
+        # due to the inplace writes.
+        cond_gm = fwd_only(cond_fn_out, fake_args, run_functional_passes=False)
+        body_gm = fwd_only(body_fn_out, fake_args, run_functional_passes=False)
```

Comment:
I guess it's kinda interesting that we're using pattern-matching infra here @eellison

Reply:
I still wonder about doing this as decomposition. Like with this pass here that [decomposes all auto functionalized operators](https://github.com/pytorch/pytorch/blob/592e3a3d4069029946ec4c8d103a313806c53a88/torch/_inductor/fx_passes/post_grad.py#L820-L918)..

if there is a direct translation from scan -> while loop it should be possible.

Reply:
Let me try this then.